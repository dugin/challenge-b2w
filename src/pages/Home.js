import React, {PureComponent} from 'react';
import styled from 'styled-components';
import {Box} from "grid-styled";
import Card from "../components/Card";
import Button from "../components/Button";
import {withRouter} from "react-router-dom";
import {connect} from "react-redux";
import {fetchPlanet, fetchPlanets} from "../planet/planet.action";

const Container = styled(Box)`
  max-width: 470px;
  margin: auto;
  text-align: center;
  position: relative;
  
  & > * {
    margin-bottom: 20px;
  }
`;


class Home extends PureComponent {

    onNext = () => {
        if (this.props.planetID) {
            this.props.history.replace(`/planet/${this.props.planetID}`);

            this.fetchPlanet(this.props.planetID);
        }
        else {
            this.fetchPlanet();
            this.props.history.replace(`/`)
        }
    };

    componentDidMount() {
        const {pathname} = this.props.location;

        if (pathname.includes('planet'))
            this.fetchPlanet(pathname.substring(pathname.lastIndexOf('/') + 1));

        else if (pathname === '/')
            this.fetchPlanet();
    }

    fetchPlanet(id) {
        this.props.dispatch(id ? fetchPlanet(id) : fetchPlanets());
    }

    render() {
        return (
            <Container
                width={[7 / 8, 4 / 7, 1 / 2, 3 / 7]}>

                <Card
                    isLoading={this.props.isLoading}
                    isError={this.props.error}
                    planet={this.props.planet}/>
                <Button
                    isLoading={this.props.isLoading}
                    text={'Next'}
                    onPress={() => this.onNext()}/>
            </Container>
        );
    }
}

const mapStateToProps = (state) => {
    const {planet, amount, isDone, error} = state.planets;

    return {
        planet,
        planetID: amount !== 0 && Math.floor((Math.random() * amount) + 1),
        isLoading: !isDone,
        error: !!error
    }
};

export default connect(mapStateToProps)(withRouter(Home));
