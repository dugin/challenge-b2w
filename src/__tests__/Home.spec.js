import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Home from "../pages/Home";

describe('App Component', () => {

    const home = shallow(<Home/>);

    it('renders properly', () => {
        expect(toJson(home)).toMatchSnapshot();
    });
});