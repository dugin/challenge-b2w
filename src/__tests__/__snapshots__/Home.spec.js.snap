// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders properly 1`] = `
<div>
  <Styled(Box)
    width={
      Array [
        0.875,
        0.5714285714285714,
        0.5,
        0.42857142857142855,
      ]
    }
  >
    <Card
      info="xyz"
    />
    <Button
      text="Next"
    />
  </Styled(Box)>
</div>
`;
